editor/plugins/camera_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,145 @@
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "camera_2d_editor_plugin.h"
+
+#include "canvas_item_editor_plugin.h"
+#include "core/config/project_settings.h"
+#include "editor/editor_node.h"
+#include "editor/themes/editor_scale.h"
+#include "scene/gui/label.h"
+#include "scene/gui/menu_button.h"
+
+void Camera2DEditor::edit(Camera2D *p_camera) {
+	if (p_camera == selected_camera) {
+		return;
+	}
+	selected_camera = p_camera;
+}
+
+void Camera2DEditor::_menu_option(int p_option) {
+	switch (p_option) {
+		case MENU_SNAP_LIMITS_TO_VIEWPORT: {
+			if (selected_camera == nullptr) {
+				break;
+			}
+			selected_camera->set_limit(SIDE_LEFT, 0);
+			selected_camera->set_limit(SIDE_TOP, 0);
+			selected_camera->set_limit(SIDE_RIGHT, GLOBAL_GET("display/window/size/viewport_width"));
+			selected_camera->set_limit(SIDE_BOTTOM, GLOBAL_GET("display/window/size/viewport_height"));
+			break;
+		}
+		default:
+			break;
+	}
+}
+
+void Camera2DEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_THEME_CHANGED: {
+			options->set_button_icon(get_editor_theme_icon(SNAME("Camera2D")));
+			options->get_popup()->set_item_icon(MENU_SNAP_LIMITS_TO_VIEWPORT, get_editor_theme_icon(SNAME("Camera2D")));
+		} break;
+	}
+}
+
+Camera2DEditor::Camera2DEditor() {
+	options = memnew(MenuButton);
+
+	CanvasItemEditor::get_singleton()->add_control_to_menu_panel(options);
+
+	options->set_text(TTR("Camera2D"));
+
+	options->get_popup()->add_item(TTR("Snap the Limits to the Viewport"), MENU_SNAP_LIMITS_TO_VIEWPORT);
+	options->set_switch_on_hover(true);
+
+	options->get_popup()->connect(SceneStringName(id_pressed), callable_mp(this, &Camera2DEditor::_menu_option));
+}
+
+void Camera2DEditorPlugin::_update_approach_text_visibility() {
+	if (camera_2d_editor->selected_camera == nullptr) {
+		print_line("Selected camera is null");
+		return;
+	}
+	approach_to_move_rect->set_visible(camera_2d_editor->selected_camera->is_limit_enabled());
+}
+
+void Camera2DEditorPlugin::edit(Object *p_object) {
+	Camera2D *cam = Object::cast_to<Camera2D>(p_object);
+	if (cam != nullptr) {
```

Comment:
This logic is flawed. You should first check for `selected_camera` and disconnect the signal from it, then connect to `cam`. You are disconnecting and connecting the same camera.

Usually the editor has its own `edit()` method that handles it, the plugin only forwards it.

Reply:
Still wrong.
If `cam` is null, you should still disconnect old camera and stop editing it.

Reply:
I may be wrong, but also instead of nullptr check is_null()/is_valid() should be used.

Reply:
`is_null()/valid()` is for RefCounted, not for Nodes.